### PR TITLE
Enable razor UTs in-memory compilation

### LINKED ISSUE #9010: Enable razor UTs in-memory compilation
- [ ] Link all the steps to allow razor UTs to run with in-memory compilation.
- [ ] Fix [razor symbol reference](https://trello.com/c/a1wUeO5j/1319-fix-razor-uts-for-symbol-references-2devs-1w) issue
- [ ] Final cleanup:
  - There's probably a build locator (or other) dependency that will go away
  - Verifier `IsRazor` property and `CompileRazor`, `LangVersion`, `PrepareRazorProject`, `PrepareRazorFiles` and `RestorePackages` methods will go away

### PR DESCRIPTION
**Pull request by CristianAmbrosini**
Fixes #9010

### CODE REVIEW

**Review by Tim-Pohlmann — CHANGES_REQUESTED** [review]
The main code changes look good. I have some comments about the tests.

File: analyzers/tests/SonarAnalyzer.Test/AnalysisContext/SonarAnalysisContextTest.Register.cs [hunk 1]
```diff
@@ -428,8 +428,8 @@ public async Task SonarCompilationStartAnalysisContext_RegisterSymbolStartAction
     public void DisabledRules_ForRazor_DoNotRaise(string ruleId, string extension) =>
         new VerifierBuilder()
             .AddAnalyzer(() => new DummyAnalyzerWithLocation(ruleId, DiagnosticDescriptorFactory.MainSourceScopeTag))
-            .WithAdditionalFilePath(AnalysisScaffolding.CreateSonarProjectConfig(TestContext, ProjectType.Product))
             .AddSnippet(Snippet(extension), $"SomeFile.{extension}")
+            .WithOptions(ParseOptionsHelper.FromCSharp9)
```
[the remark this answers is not in the document]
Comment (CristianAmbrosini) [reply analyzers/tests/SonarAnalyzer.Test/AnalysisContext/SonarAnalysisContextTest.Register.cs:432]: Not needed indeed

File: analyzers/tests/SonarAnalyzer.Test/Rules/UnnecessaryUsingsTest.cs [hunk 2]
```diff
@@ -53,42 +53,34 @@ public class UnnecessaryUsingsTest
...
         [DataTestMethod]
         [DataRow("_Imports.razor")]
         [DataRow("_imports.razor")]
         [DataRow("_iMpoRts.razor")]
         public void UnnecessaryUsings_RazorImportsRazorFile_NoIssueReported(string fileName) =>
-            builder
-                .AddSnippet(@"@using System.Text.Json;", fileName)
-                .WithAdditionalFilePath(AnalysisScaffolding.CreateSonarProjectConfig(TestContext, ProjectType.Product))
-                .VerifyNoIssueReported();
+            builder.AddSnippet(@"@using System.Text.Json;", fileName).VerifyNoIssueReported(); // ToDo TAKE A LOOK
```
[the remark this answers is not in the document]
Comment (CristianAmbrosini) [reply analyzers/tests/SonarAnalyzer.Test/Rules/UnnecessaryUsingsTest.cs:63]: Yeah, we added this comment for some reason, but I don't remember why. @sebastien-marichal, do you have any idea?

[the remark this answers is not in the document]
Comment (sebastien-marichal) [reply analyzers/tests/SonarAnalyzer.Test/Rules/UnnecessaryUsingsTest.cs:63]: My guess would be that we saw the test is green even though we don't provide the `System.Text.Json` reference. Which seems to be needed in the next test.

It probably should have a compilation error or we don't need to provide the reference in the next test.

File: analyzers/tests/SonarAnalyzer.TestFramework/Dependencies/Microsoft.CodeAnalysis.Razor.Compiler.SourceGenerators.deps.json
[the remark this answers is not in the document]
Comment (CristianAmbrosini) [reply analyzers/tests/SonarAnalyzer.TestFramework/Dependencies/Microsoft.CodeAnalysis.Razor.Compiler.SourceGenerators.deps.json:1]: let's keep it there, it's a temporary solution anyway, the entire folder will be removed in the follow up PR

File: analyzers/tests/SonarAnalyzer.TestFramework/Dependencies/RazorSourceGenerator.razorencconfig
[the remark this answers is not in the document]
Comment (CristianAmbrosini) [reply analyzers/tests/SonarAnalyzer.TestFramework/Dependencies/RazorSourceGenerator.razorencconfig:1]: Same as above

File: analyzers/tests/SonarAnalyzer.Test/Rules/UnnecessaryUsingsTest.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.Test/Rules/UnnecessaryUsingsTest.cs]: The difference lies in `Verify` vs. `VerifyNoIssueReported`. I'm not sure why exactly, but the test is fine as is.

**Review by Tim-Pohlmann — APPROVED** [review]
LGTM!

**Pull request merged**